autode/bracket/base.py

```diff
@@ -0,0 +1,284 @@
+
+    @property
+    @abstractmethod
+    def _micro_iter(self) -> int:
+        """Total number of micro-iterations run with this method"""
+
+    @abstractmethod
+    def _initialise_run(self) -> None:
+        """Initialise the bracketing method run"""
+
+    @abstractmethod
+    def _step(self) -> None:
+        """
+        One step of the bracket method, with one macro-iteration
+        and multiple micro-iterations. This must also set new
+        coordinates for the next step
+        """
+
+    def _log_convergence(self) -> None:
+        """
+        Log the convergence of the bracket method. Only logs macro-iters,
+        subclasses may implement further logging for micro-iters
+        """
+        logger.info(
+            f"Macro-iteration #{self._macro_iter}: Distance = "
+            f"{self.imgpair.dist:.4f}; Energy (initial species) = "
+            f"{self.imgpair.left_coord.e:.6f}; Energy (final species) = "
+            f"{self.imgpair.right_coord.e:.6f}"
+        )
+
+    @property
+    def _exceeded_maximum_iteration(self) -> bool:
+        """Whether it has exceeded the number of maximum micro-iterations"""
+        if self._micro_iter >= self._maxiter:
+            return True
+        else:
+            return False
+
+    def calculate(
+        self,
+        method: "Method",
+        n_cores: int,
+    ) -> None:
+        """
+        Run the bracketing method calculation using the method for
+        energy/gradient calculation, with n_cores. Runs CI-NEB at
+        the end if requested
+        """
+        self.imgpair.set_method_and_n_cores(method, n_cores)
+        self._initialise_run()
+
+        logger.info(
+            f"Starting {self._method_name} method to find transition state"
+        )
+
+        while not self.converged:
+            self._step()
+
+            if self.imgpair.has_jumped_over_barrier():
+                logger.warning(
+                    "One image has probably jumped over the barrier, in"
+                    f" {self._method_name} TS search. Please check the"
+                    f" results carefully"
+                )
+                if self._barrier_check:
+                    logger.info(f"Stopping {self._method_name} calculation")
+                    break
+
+            if self._exceeded_maximum_iteration:
+                break
+
+            self._log_convergence()
+
+        # exited main loop, run CI-NEB if required and bracket converged
+        if self._should_run_cineb and self.converged:
+            self.run_cineb()
+
+        # print final message
+        logger.info(
+            f"Finished {self._method_name} procedure in {self._macro_iter} "
+            f"macro-iterations consisting of {self._micro_iter} micro-"
+            f"iterations (optimiser steps). {self._method_name} is "
+            f"{'converged' if self.converged else 'not converged'}"
+        )
+        return None
+
+    def run(self) -> None:
```

Comment:
@t-young31 I wanted to have a method without any user input (sort of an easy thing to run without worrying about anything), but maybe the run() is unnecessary, because I already have the calculate method. I was trying to copy the interface of NEB classes.